File: backend/src/test/java/com/codeloom/backend/it/TestAuthentication.java
package com.codeloom.backend.it;

import com.codeloom.backend.security.UserRole;
import java.util.*;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;

final class TestAuthentication {
  static final UUID TEST_USER_ID = UUID.fromString("00000000-0000-0000-0000-000000000001");

  private TestAuthentication() {}

  static JwtAuthenticationToken user(UserRole... roles) {
    Jwt jwt =
        Jwt.withTokenValue("test-token")
            .header("alg", "none")
            .subject(TEST_USER_ID.toString())
            .build();
    return new JwtAuthenticationToken(
        jwt, Arrays.stream(roles).map(r -> new SimpleGrantedAuthority(r.getRoleName())).toList());
  }

  static JwtAuthenticationToken admin() {
    return user(UserRole.USER, UserRole.ADMIN);
  }
}

File: backend/src/test/kotlin/com/codeloom/backend/it/TestAuthentication.kt
package com.codeloom.backend.it

import com.codeloom.backend.security.UserRole
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.test.web.servlet.MockHttpServletRequestDsl
import java.util.*

val TEST_USER_ID: UUID = UUID.fromString("00000000-0000-0000-0000-000000000001")

fun MockHttpServletRequestDsl.initUser(
    userId: UUID = TEST_USER_ID,
    vararg roles: UserRole = arrayOf(UserRole.USER, UserRole.ADMIN),
) {
    val jwt =
        Jwt.withTokenValue("test-token")
            .header("alg", "none")
            .subject(userId.toString())
            .build()
    principal =
        JwtAuthenticationToken(
            jwt,
            roles.map { role -> SimpleGrantedAuthority(role.roleName) },
        )
}
